src/async-directive.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright (c) 2021 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Directive, Part, PartInfo} from './directive.js';
+import * as legacyLit from './lit-html.js';
+
+/**
+ * A superclass for directives that need to asynchronously update.
+ */
+export abstract class AsyncDirective extends Directive {
+  private readonly ddPart: legacyLit.Part;
+  private renderedYet = false;
+
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    this.ddPart = (partInfo as Part).legacyPart;
+  }
+
+  private ddGetNode(): Node|undefined {
+    if (this.ddPart instanceof legacyLit.NodePart) {
+      return this.ddPart.startNode;
+    } else if (this.ddPart instanceof legacyLit.EventPart) {
+      return this.ddPart.element;
+    } else if (this.ddPart instanceof legacyLit.BooleanAttributePart) {
+      return this.ddPart.element;
+    } else if (
+        this.ddPart instanceof legacyLit.PropertyPart ||
+        this.ddPart instanceof legacyLit.AttributePart) {
+      return this.ddPart.committer.element;
+    }
+    return undefined;
+  }
+
+  private shouldRender() {
+    if (!this.renderedYet) {
+      this.renderedYet = true;
+      return true;
+    }
+    const node = this.ddGetNode();
+    if (node != null && !node.isConnected) {
+      // node is disconnected, do not render
+      return false;
+    }
+    return true;
+  }
+
+  setValue(value: unknown) {
+    if (!this.shouldRender()) {
+      // node is disconnected, do nothing.
+      return;
+    }
+
+    this.ddPart.setValue(value);
+    this.ddPart.commit();
+  }
+
+  /**
+   * User callbacks for implementing logic to release any
+   * resources/subscriptions that may have been retained by this directive.
+   * Since directives may also be re-connected, `reconnected` should also be
+   * implemented to restore the working state of the directive prior to the next
+   * render.
+   *
+   * In the v1 version of these APIs, we don't monitor for disconnection and
+   * reconnection, we only call these methods when setValue is called.
```

Comment:
Docs seem wrong here... they're never called right?

Reply:
@rictic

Reply:
They won't get called in V1 (we should just document that fact in the API docs). The idea is this is ok because a V1 directive couldn't have relied on the callback being called.

The only reason to put them here on the interface is so that users can go ahead and write code that _will_ have the callbacks run in V2.

Reply:
Yeah, sorry, this doc is out of date. I think for a little while I have a halfway implementation here, but it diverged enough from the behavior of the v2 version that it seemed like it might do more harm than good.

+1 to having clear docs that these are just for interface compatibility, that lit-html v1 will not call them